### PR TITLE
cli: add --hard flag to rollback command to remove block as well 

### PR DESCRIPTION
**Pull request by pirtleshell**
Cherrypick / backport of https://github.com/cometbft/cometbft/commit/bd9ed68a0483f38afee88bbda0f3b41757a6d21f

Before this commit, `RollbackState` (used only by [the `rollback` CLI command](https://github.com/Kava-Labs/cosmos-sdk/blob/b8317d86cd48706cb0ec5e55bf83805999803386/server/rollback.go#L35) of cosmos-sdk) rolled back the tendermint state by one version, but not the block state.

This means that the command can only be used to rollback one block, because if the most recent tendermint store version (block number) matches the most recent block store version, the rollback operation is a no-op. ([code](https://github.com/Kava-Labs/cometbft/blob/277b0853db3f67865a55aa1c54f59790b5f591be/state/rollback.go#L33-L38).

The commit updates the `RollbackState` method to have a `removeBlock` argument that is used to also trigger the removal of the block store version, allowing for rolling back the state of multiple blocks.

Once merged, it will require a patch to cosmos-sdk to handle the new function signature of `RollbackState` in the rollback CLI command.

tl;dr: `RollbackState` can now be used to rollback a sequence of blocks, not just the most recent block. This is not consensus breaking as the method is only used in the CLI.

### CODE REVIEW

**Review by drklee3 — CHANGES_REQUESTED** [review]
Looks good, just some errors with a mock and rollback_test that need to be updated

```
# github.com/tendermint/tendermint/rpc/core [github.com/tendermint/tendermint/rpc/core.test]
rpc/core/blocks_test.go:89:19: cannot use mockBlockStore{…} (value of type mockBlockStore) as "github.com/tendermint/tendermint/state".BlockStore value in assignment: mockBlockStore does not implement "github.com/tendermint/tendermint/state".BlockStore (missing method DeleteLatestBlock)
FAIL	github.com/tendermint/tendermint/rpc/core [build failed]
ok  	github.com/tendermint/tendermint/rpc/core/types	0.012s
ok  	github.com/tendermint/tendermint/rpc/grpc	0.360s
ok  	github.com/tendermint/tendermint/rpc/jsonrpc	8.033s
ok  	github.com/tendermint/tendermint/rpc/jsonrpc/client	9.005s
ok  	github.com/tendermint/tendermint/rpc/jsonrpc/server	0.138s
?   	github.com/tendermint/tendermint/rpc/jsonrpc/test	[no test files]
ok  	github.com/tendermint/tendermint/rpc/jsonrpc/types	0.008s
?   	github.com/tendermint/tendermint/rpc/test	[no test files]
?   	github.com/tendermint/tendermint/scripts/json2wal	[no test files]
?   	github.com/tendermint/tendermint/scripts/wal2json	[no test files]
# github.com/tendermint/tendermint/state_test [github.com/tendermint/tendermint/state.test]
state/rollback_test.go:96:17: params.Version.App undefined (type "github.com/tendermint/tendermint/proto/tendermint/types".VersionParams has no field or method App)
state/rollback_test.go:101:24: undefined: tmversion
state/rollback_test.go:111:31: params.Hash undefined (type *"github.com/tendermint/tendermint/proto/tendermint/types".ConsensusParams has no field or method Hash)
state/rollback_test.go:119:18: assignment mismatch: 2 variables but block.MakePartSet returns 1 value
state/rollback_test.go:124:12: undefined: tmstate
state/rollback_test.go:143:24: undefined: tmversion
state/rollback_test.go:153:31: params.Hash undefined (type *"github.com/tendermint/tendermint/proto/tendermint/types".ConsensusParams has no field or method Hash)
state/rollback_test.go:161:22: assignment mismatch: 2 variables but nextBlock.MakePartSet returns 1 value
state/rollback_test.go:178:17: params.Version.App undefined (type "github.com/tendermint/tendermint/proto/tendermint/types".VersionParams has no field or method App)
state/rollback_test.go:181:12: undefined: tmstate
state/rollback_test.go:181:12: too many errors
FAIL	github.com/tendermint/tendermint/state [build failed]
```

**Review by drklee3 — APPROVED** [review]
